### ISSUE TITLE
[Sandbox] Substation

### ISSUE THREAD

**Issue by jshlbrd** [cncf/sandbox#117]
### Application contact emails

jliburdi@brex.com

### Project Summary

Substation is a toolkit for routing, normalizing, and enriching security event and audit logs.

### Project Description

Substation provides three core capabilities:
- Deploy cloud native, serverless data pipelines in minutes
- Evaluate and transform event logs in real-time
- Create custom data processing applications written in Go

Substation is inspired by older data pipeline systems like Logstash and Fluentd, but takes a cloud native approach to deployment and maintenance by deploying directly on cloud services (such as AWS Lambda) with a focus on being highly configurable by end users with features not seen in other systems ([robust condition logic](https://substation.readme.io/docs/conditions), support for [external key-value stores](https://substation.readme.io/docs/key-value-stores), and [configurations as code](https://substation.readme.io/docs/configurations-as-code) to name a few).

### Org repo URL (provide if all repos under the org are in scope of the application)

N/A

### Project repo URL in scope of application

https://github.com/brexhq/substation

### Website URL

https://substation.readme.io/

### Roadmap

https://github.com/brexhq/substation/issues

### Roadmap context

The project uses GitHub issues for tracking roadmap progress. In 2022 the project had [defined a 2023 roadmap](https://github.com/brexhq/substation/discussions/45) that was implemented by mid 2024. As of now there is no defined roadmap for 2024 (or beyond), but with acceptance to CNCF we can revisit this. The project strictly follows SemVer 2.0 and produces [iterative releases](https://github.com/brexhq/substation/releases) (up to once per week) and we also have open discussions about future releases ([v1](https://github.com/brexhq/substation/discussions/127), [v2](https://github.com/brexhq/substation/discussions/224)).

### Contributing Guide

https://github.com/brexhq/substation/blob/main/CONTRIBUTING.md

### Code of Conduct (CoC)

https://github.com/brexhq/substation/blob/main/CODE_OF_CONDUCT.md

### Adopters

https://github.com/brexhq/substation/blob/main/ADOPTERS.md

### Contributing or Sponsoring Org

https://www.brex.com/

### Maintainers file

https://github.com/brexhq/substation/blob/main/CODEOWNERS

### IP Policy

- [X] If the project is accepted, I agree the project will follow the CNCF IP Policy

### Trademark and accounts

- [X] If the project is accepted, I agree to donate all project trademarks and accounts to the CNCF

### Why CNCF?

The CNCF can provide the project with more exposure to future contributors and cross-organization governance (beyond Brex). There are a number of features that would be beneficial to add to the project (such as support for additional cloud service providers, like GCP or Azure, and out of the box data transform configurations, such as AWS CloudTrail), but the team at Brex can't dedicate time to these additional features. 

Without additional support the project will continue to be focused on providing core capabilities built for AWS services.

### Benefit to the Landscape

Cloud native event log routing, filtering, and normalization is a challenge faced by many organizations; and interestingly, many organizations that I speak with aren't familiar with the current "state of the art" in this space and are relying on vendors to solve the problem. Today CNCF has one project that partially solves similar problems (Fluentd, mentioned in more detail below), but I think there is room for more than one solution in the CNCF, especially given the differences between these two projects.

Brex has been successfully using Substation in production for more than three years to reduce cost and improve the usefulness of our event logs (and are aware of a few other companies using it for the same use case), more information can be gleaned from blog posts and presentations over the years:
- https://medium.com/brexeng/announcing-substation-188d049d979b
- https://medium.com/brexeng/releasing-substation-v1-0-4d0314cbc45b
- https://www.youtube.com/watch?v=im5xfAHeiRo

This is a "production-grade" project that has data processing functions not seen in other tools (mentioned above) and that was built with reliability, maintenance, and low-costs as built-in features.

### Cloud Native 'Fit'

The project supports many cloud native features and, out of the box, is intended to be deployed on AWS as serverless functions connected to serverless services. This includes:
- Deploying compute on Lambda with service support for Kinesis, DynamoDB, S3 (and many more).
- Deploying as containers (on Lambda, but can also be built for orchestration platforms).
- Designed to run anywhere -- with some effort, new Go applications that utilize the Substation package can be written to run on any cloud service provider.

### Cloud Native 'Integration'

Substation does not have any dependencies on existing CNCF projects, but it can complement projects like Cilium and Falco by acting as an event log router, filter, and normalization system for the logs they produce. It can also format audit logs according to the [OpenTelemetry SemConv](https://opentelemetry.io/docs/specs/semconv/) standard.

### Cloud Native Overlap

This project overlaps with [Fluentd](https://www.cncf.io/projects/fluentd/) in the sense that both projects can solve similar problems (event log routing and filtering). Where the two projects differ is in their underlying design (Substation is a package written in Go with out of the box support for AWS serverless cloud deployments, and can be extended for deployment in the cloud, on-prem, or locally; Fluentd is an application written in Ruby with out of the box support for deploying on Kubernetes). 

Substation is more of a "toolkit" that can be used to implement an event log routing and filtering solution, among other solutions (more info [here](https://substation.readme.io/docs/use-cases)). Substation was primarily built for security use cases, which is often a superset of observability use cases due to their increased complexity, and was designed for ease of deployment and maintenance ("zero ops").

### Similar projects

- https://www.fluentd.org/ (CNCF)
- https://www.elastic.co/logstash
- https://github.com/redpanda-data/benthos

### Landscape

No.

### Business Product or Service to Project separation

N/A

### Project presentations

- TAG Security (9/18/2024): https://github.com/cncf/tag-security/issues/1356#issuecomment-2359844921

### Project champions

N/A

**Comment by angellk** [comment]
@jshlbrd please [submit a presentation issue](https://github.com/cncf/tag-security/issues/new?assignees=&labels=usecase-presentation%2C+triage-required&template=presentation.md&title=%5BPresentation%5D+Presentation+Title) to give an overview of the project to [TAG Security](https://github.com/cncf/tag-security/)

**Comment by jshlbrd** [comment]
> @jshlbrd please [submit a presentation issue](https://github.com/cncf/tag-security/issues/new?assignees=&labels=usecase-presentation%2C+triage-required&template=presentation.md&title=%5BPresentation%5D+Presentation+Title) to give an overview of the project to [TAG Security](https://github.com/cncf/tag-security/)

Thanks @angellk, I submitted to present!

**Comment by jshlbrd** [comment]
Updates made to the project based on the TAG Security feedback:
- Added CodeQL for static analysis -- it identified some high risk but low impact findings that we'll be addressing.
- Added OpenSSF automation and badge -- [currently rated at 7.1](https://securityscorecards.dev/viewer/?uri=github.com/brexhq/substation) and will go up as findings are addressed.
- Adding fuzzing to exported code / public API.

**Comment by rochaporto** [comment]
One thing to consider is the project currently has a MIT License.

https://github.com/brexhq/substation/blob/main/LICENSE
https://github.com/cncf/foundation/blob/main/charter.md#11-ip-policy

**Comment by jshlbrd** [comment]
> [@jshlbrd](https://github.com/jshlbrd) can you add the pointer to the relevant meeting notes / recording with the TAG-Security feedback?

Yep! https://github.com/cncf/tag-security/issues/1356#issuecomment-2359844921

**Comment by jberkus** [comment]
TAG Contributor strategy has reviewed this project and found the following:

* The [contributor guide](https://github.com/brexhq/substation/blob/main/CONTRIBUTING.md) is somewhat more detailed than average, but is missing information about PR review and approval.
* There is no written governance that we could find.
* There is no public roadmap.
* Maintainer information is not public (CODEOWNERS links to a Github group, the membership of which is confidential)

This review is for the TOC’s information only.  Sandbox projects are not required to have full governance or contributor documentation.

**Comment by halcyondude** [comment]
I've read through the ([substation documentation](https://substation.readme.io)) and I have a few concerns about admitting the project to the CNCF Sandbox.

## Substation Summary

Substation uses the flexible and capable [jsonnet](https://jsonnet.org/) language to provide a rich library of transformations, mutations, and advanced filtering functions which are useful for data processing, log routing/shipping, and for 1:N broadcasting of events to systems in parallel (e.g. production, data lake, SEIM). However, it does so exclusively on AWS, doesn't integrate with open standards such as OpenTelemetry, and doesn't enjoy sufficient community contribution nor project governance for admission into the CNCF Sandbox.

### On Substation's dependencies

It takes notable hard dependencies on:

* AWS exclusively. There is not a local option implemented, only a CLI which can be run locally.
* Terraform is used for AWS Infra mgmt.
* [GJSON](https://github.com/tidwall/gjson) and [SJSON](https://github.com/tidwall/sjson).  GJSON syntax is rich, expressive, and interesting. To use Substation is to adopt this syntax for handling JSON. It's also MIT Licensed, not a broadly or widely adopted standard, and is governed and controlled only by its author ([tidwall](https://github.com/tidwall)).
* Jsonnet

### On Jsonnet

While jsonnet is elegant to some (myself included), it is also characterized by a [somewhat steep learning curve](https://jsonnet.org/ref/language.html), and requires a willingness to embrace its [Design Rationale](https://jsonnet.org/articles/design.html) to be productive. ["Jsonnet is a purely functional language w/ OO features"](https://jsonnet.org/ref/language.html) and takes a different mindset and approach than most other languages uses in the CNCF's project portfolio. See the Jsonnet documentation for an [article which elaborates](https://jsonnet.org/articles/comparisons.html) on how it compares to other languages.

### Substation is NOT cloud native.

Substation, as implemented, runs exclusively on a single cloud (AWS). There does not appear to be an abstraction layer that would allow incremental support for additional clouds or running the project locally.
 
_from https://substation.readme.io/docs/architecture-design:_
![Image](https://github.com/user-attachments/assets/c431f785-8d54-47ff-9271-6da5f440fc86)

> _from https://substation.readme.io/docs/app-aws-substation:_
>
> [API Gateway](https://docs.aws.amazon.com/lambda/latest/dg/services-apigateway.html) (AWS_API_GATEWAY)
> [DynamoDB (Stream)](https://docs.aws.amazon.com/lambda/latest/dg/with-ddb.html) (AWS_DYNAMODB_STREAM)
> [Kinesis Data Firehose](https://docs.aws.amazon.com/firehose/latest/dev/data-transformation.html) (AWS_KINESIS_DATA_FIREHOSE)
> [Kinesis Data Streams](https://docs.aws.amazon.com/lambda/latest/dg/with-kinesis.html) (AWS_KINESIS_DATA_STREAM)
> [Lambda (Asynchronous Invocation)](https://docs.aws.amazon.com/lambda/latest/dg/invocation-async.html) (AWS_LAMBDA)
> [Lambda (Synchronous Invocation)](https://docs.aws.amazon.com/lambda/latest/dg/invocation-sync.html) (AWS_LAMBDA)
> [S3](https://docs.aws.amazon.com/lambda/latest/dg/with-s3.html) (AWS_S3)
> [S3 via SNS](https://docs.aws.amazon.com/AmazonS3/latest/userguide/ways-to-add-notification-config-to-bucket.html) (AWS_S3_SNS)
> [SNS](https://docs.aws.amazon.com/lambda/latest/dg/with-sns.html) (AWS_SNS)
> [SQS](https://docs.aws.amazon.com/lambda/latest/dg/with-sqs.html) (AWS_SQS)

### Regarding [Substation's handling of Metrics](https://substation.readme.io/docs/metrics-destinations),

Substation supports reporting only via [AWS Embedded Metrics Format](https://docs.aws.amazon.com/AmazonCloudWatch/latest/monitoring/CloudWatch_Embedded_Metric_Format_Specification.html#CloudWatch_Embedded_Metric_Format_Specification_structure) to AWS's CloudWatch service, from the AWS Lambda it's running from. There is neither support nor roadmap for open standards such as OpenTelemetry.

### Regarding Substation's contributor community

Substation appears to be largely authored by [jshlbrd](https://github.com/jshlbrd), with additional contribution from [shellcromancer](https://github.com/shellcromancer) (22 commits) and [Mallika05](https://github.com/Mallika05) (7 commits), and a few other contributors (3, 4, 2, 2, 1 commits each). The project is controlled and maintained presently by a single company (Brex).

<img width="848" alt="Image" src="https://github.com/user-attachments/assets/dd81f607-701a-4b87-94c5-caa81b2f0e9c" />

**Comment by jshlbrd** [comment]
@halcyondude Thanks for the review, there's an aspect to this project that I think was overlooked. You reference it here:

> Substation, as implemented, runs exclusively on a single cloud (AWS). There does not appear to be an abstraction layer that would allow incremental support for additional clouds or running the project locally.

The [package is the abstraction layer](https://github.com/brexhq/substation/blob/main/substation.go) and new applications can be written for additional cloud environments, on-premise, or local deployment. Everything under `cmd/` is an implementation of the abstraction layer for specific use cases (ex. AWS). The team at Brex runs on AWS, so there is no out of the box support for other clouds or deployment models, but that does not mean they cannot exist in the future. This is mentioned in the application under **Why CNCF**.

The "hard" dependencies you mentioned are also "soft" dependencies (except for GJSON and SJSON, which are implemented in the abstraction layer). For example, the project ships with Terraform support, but if someone wants to deploy with CloudFormation instead, then they can do that if they want to put in the effort to create the templates. 

Hope that helps! Happy to answer any other questions about these topics.

**Comment by jshlbrd** [comment]
Hey all, what's the status of this submission? I see it's in a postponed state on the project board, but there's no next steps on how to move forward or not. Thanks!

**Comment by andrew-kline** [comment]
Howdy, wanted to follow up here on Josh's most recent ask - are there any next steps we can take?
